package/metrics_daemon/metrics_daemon/src/metrics_daemon.c

```diff
@@ -0,0 +1,378 @@
+                                                    const char *root,
+                                                    unsigned int root_len)
+{
+  char ch = '/';
+  const char *pdest = NULL;
+  bool found_root = false;
+  pdest = strchr(process_path, ch); // get the first slash position
+  unsigned int str_len = (unsigned int)strlen(process_path);
+  if (jobj_root == NULL || pdest < process_path) {
+    return jobj_root;
+  }
+  struct json_object *json_current = jobj_root;
+  while (pdest != NULL) {
+    pdest++;
+    if (pdest > process_path + str_len) return json_current;
+    if (pdest[0] == '.') // skip the '.' on the file name
+    {
+      pdest++;
+    }
+    if (pdest > process_path + str_len) return json_current;
+    char *pdest2 = strchr(pdest, ch); // search the second slash
+    if (pdest2 != NULL) {
+      unsigned int strlen;
+      if (pdest2 <= pdest) // defensive code
+        return json_current;
+      else
+        strlen = (unsigned)(pdest2 - pdest); // cast to unsigned since the negative and zero
+                                             // case is handled by the if statement
+      if (!found_root && strncmp(pdest, root, root_len) == 0) // search the root first
+      {
+        found_root = true;
+        pdest--;
+        continue;
+      }
+      if (found_root) {
+        bool found_target = false;
+        json_object_object_foreach(json_current, key, val)
+        { // for each subfolder, search the name as key in json tree
+          if (strncmp((pdest), key, strlen - 1) == 0) // if find the folder name, continue on the
+                                                      // loop with the new current node
+          {
+            json_current = val;
+            found_target = true;
+            break;
+          }
+        }
+        if (!found_target) // if couldn't find target on the tree, return
+                           // current node to add new node
+          return json_current;
+      }
+    }
+    pdest = pdest2; // if there is still sub-folders, update the pdest to pdest2
+  }
+  return json_current;
+}
+
+
+/**
+ * @brief function get called by ftw for each of the sub-folder or file under
+ * the path
+ *
+ * @param fpath  file path in ASCII
+ * @param sb     file status struct
+ * @param tflag  flag indicates file/folder type
+ * @return integer that indicates if the walk through failed.
+ */
+int handle_walk_path(const char *fpath, const struct stat *sb, int tflag)
+{
+
+  if (first_folder == true) // skip the root since it is handled by main()
+  {
+    first_folder = false;
+    return 0;
+  }
+  char *bname = extract_filename(fpath);
+  if (bname == NULL) {
+    piksi_log(LOG_ERR, "FTW: invalid file name\n"); // For now, only support file and forlder
+    return -1;
+  }
+  json_object *jobj_lc_node =
+    loop_through_folder_name(fpath,
+                             root_name,
+                             root_length); // get the least common node between json object and file
```

Comment:
Maybe these comments can be moved above the call in order to avoid the strange formatting results?